library/ajax/username_report_ajax.php

```diff
@@ -0,0 +1,306 @@
+            return $session;
+
+        }
+    }
+
+
+    //if the next event is a logout, we either return the timeout or the difference to the next
+    //logout, which ever is shorter.
+    if($nextEvent == 'logout'){
+
+        if(($logOutDateTime - $logInDateTime) > $sessionTimeout){
+
+            $session = round(($sessionTimeout) / (60 * 60), 2);
+            return $session;
+
+        }else{
+
+            $session = round(($logOutDateTime - $logInDateTime) / (60 * 60), 2);
+            return $session;
+        }
+    }
+
+    //There are events that exist.  We get the last activity before the next login.
+    //This can be a logout or an activity.
+
+    if($lastSessionActivity > 0  && (($lastSessionActivity - $logInDateTime) < $sessionTimeout)){
+
+       $session = round(($lastSessionActivity - $logInDateTime) / (60 * 60), 2);
+        return $session;
+
+    }else {
+
+        if($logInDateTime == strtotime($maxLogin)){
+            return 0;
+        }else return round(($sessionTimeout) / (60 * 60), 2);
+    }
+
+    return 0;
+}
+
+//takes in the user, and date range and returns the sum of sessions
+function getSumSessionsByUser($user, $from_date = '', $to_date = ''){
+    $sum = 0;
+    //get the list of dates and put it in an array
+    $query = "select date from log where event ='login' and user = ? ";
+        if($to_date != '')
+            $query .= " and date >= ? ";
+        if($from_date != '')
+            $query .= " and date <= DATE_ADD( ? , INTERVAL 1 DAY) ";
+
+        $result = sqlStatement($query, array($user, $from_date, $to_date));
+
+    //iterate through each of the dates. get the session time, and sum them
+    $oldDate = '';
+    $newDate = '';
+    while ($row = sqlFetchArray($result)) {
+        $newDate = $row['date'];
+        if($newDate === $oldDate)continue;
+        $sum = $sum + getSessionTime($user, $row['date']);
+        $oldDate = $newDate;
+    }
+
+    //return the sum
+
+    return $sum;
+}
+
+
+
+
+
+
+
+
+//To list out records with given criteria
+//To list out records with given criteria
+if($_POST['func']=="show_session_sums") {
+    $response['data'] = array();
+    $user_array = getUsersArray();
+    $index = 0;
+    foreach($user_array as $user){
+        $row = sqlQuery( "Select username, lname, fname from users where username = '$user'");
+        $row['sum'] = getSumSessionsByUser($user, $from_date, $to_date);
+        $row['date'] = "Date Range: " . date($DateFormat, strtotime($from_date)) . " - " . date($DateFormat, strtotime($to_date));
+        array_push($response['data'], $row);
+    }
+    echo json_encode($response);
+
+}
+
+if($_POST['func']=="show_session_times")
+{
+    $response['data'] = array();
+
+
+    $qstring = "Select  date, event, username as user, fname, lname  from log join users on username = user where event like '%login%' and event not like '%attempt%' and users.active = 1";
+
```

Comment:
@growlingflea For performance, maybe you can get all the data in one query rather than calling multiple queries in a loop?